lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -535,6 +535,12 @@ instance LiftHandler ErrEstimateTxFee where
     handler = \case
         ErrEstimateTxFeeNoSuchWallet e -> handler e
         ErrEstimateTxFeeCoinSelection e -> handler e
+        ErrEstimateTxFeeBackend _ ->
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "I can't estimate transaction fee because either"
+                , " transaction contains at least one payment output of value 0 (Byron)"
+                , " or a transaction has more than 255 inputs or outputs (Jormungandr)."
```

Comment:
Hmmm. It'd be best not to mention Byron or Jormungandr in the core code. Here, it'd be fine to simply say that the transaction is invalid and give a few examples of invalidity but, with a more general tone. Something like

"It could be that the transaction has too many inputs and/or outputs or, null outputs"

Reply:
Wouldn't a user like to know the exact reason the tx was invalid? Why not have `ErrMkStdTx Text` where `Text` is a user-presentable error with the exact failure-reason?

Reply:
Wouldn't a user like to know the exact reason the tx was invalid? Why not have `ErrMkStdTx Text` where `Text` is a user-presentable error with the exact failure-reason?

As a second-best option I think the diff is better than "It could be that the transaction has too many inputs and/or outputs or, null outputs".

Reply:
done, although separated messages fro `ErrInvalidTxOutAmount` and `ErrExceededInpsOrOuts` of `ErrValidateSelection`

Reply:
From the user-perspective I think this is good.

But `ErrInvalidTxOutAmount` and `ErrExceededInpsOrOuts` are very intertwined with the specific targets.

I'd be intrigued to let the targets themselves render the error to the user. **E.g**. by using `ErrEstimateTxFeeBackend Text`.

Something like:
<details>

```patch
diff --git a/lib/core/src/Cardano/Wallet.hs b/lib/core/src/Cardano/Wallet.hs
index 4372a1d5..cb4de208 100644
--- a/lib/core/src/Cardano/Wallet.hs
+++ b/lib/core/src/Cardano/Wallet.hs
@@ -32,7 +32,6 @@ module Cardano.Wallet
     , ErrWithRootKey (..)
     , ErrWrongPassphrase (..)
     , ErrMkStdTx (..)
-    , ErrValidateSelection (..)
     , ErrPostTx (..)
     , ErrNetworkUnreachable (..)
     , ErrCoinSelection (..)
@@ -120,7 +119,7 @@ import Cardano.Wallet.Primitive.Types
     , slotRatio
     )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..), ErrValidateSelection (..), TransactionLayer (..) )
+    ( ErrMkStdTx (..), TransactionLayer (..) )
 import Cardano.Wallet.Unsafe
     ( unsafeRunExceptT )
 import Control.Arrow
@@ -294,7 +293,7 @@ data ErrCreateUnsignedTx
 data ErrEstimateTxFee
     = ErrEstimateTxFeeNoSuchWallet ErrNoSuchWallet
     | ErrEstimateTxFeeCoinSelection ErrCoinSelection
-    | ErrEstimateTxFeeBackend ErrValidateSelection
+    | ErrEstimateTxFeeBackend Text
     deriving (Show, Eq)

 -- | Errors occuring when signing a transaction
diff --git a/lib/core/src/Cardano/Wallet/Api/Server.hs b/lib/core/src/Cardano/Wallet/Api/Server.hs
index 5731a4c8..fdfa667c 100644
--- a/lib/core/src/Cardano/Wallet/Api/Server.hs
+++ b/lib/core/src/Cardano/Wallet/Api/Server.hs
@@ -37,7 +37,6 @@ import Cardano.Wallet
     , ErrSignTx (..)
     , ErrSubmitTx (..)
     , ErrUpdatePassphrase (..)
-    , ErrValidateSelection (..)
     , ErrWalletAlreadyExists (..)
     , ErrWithRootKey (..)
     , ErrWrongPassphrase (..)
@@ -536,15 +535,10 @@ instance LiftHandler ErrEstimateTxFee where
     handler = \case
         ErrEstimateTxFeeNoSuchWallet e -> handler e
         ErrEstimateTxFeeCoinSelection e -> handler e
-        ErrEstimateTxFeeBackend ErrInvalidTxOutAmount ->
+        ErrEstimateTxFeeBackend reason ->
             apiError err403 CannotValidateSelection $ mconcat
-                [ "I can't estimate transaction fee because transaction contains"
-                , " at least one payment output of value 0."
-                ]
-        ErrEstimateTxFeeBackend ErrExceededInpsOrOuts ->
-            apiError err403 CannotValidateSelection $ mconcat
-                [ "I can't estimate transaction fee because transaction has either"
-                , " more than 255 inputs or more than 255 outputs."
+                [ "I can't estimate transaction fee because: "
+                , reason
                 ]

 instance LiftHandler ErrSignTx where
diff --git a/lib/core/src/Cardano/Wallet/Transaction.hs b/lib/core/src/Cardano/Wallet/Transaction.hs
index 92ae1365..f846cbfe 100644
--- a/lib/core/src/Cardano/Wallet/Transaction.hs
+++ b/lib/core/src/Cardano/Wallet/Transaction.hs
@@ -18,7 +18,6 @@ module Cardano.Wallet.Transaction

     -- * Errors
     , ErrMkStdTx (..)
-    , ErrValidateSelection (..)

     -- * Backend helpers
     , estimateMaxNumberOfInputsBase
@@ -35,6 +34,8 @@ import Cardano.Wallet.Primitive.Types
     ( Address (..), Hash (..), Tx, TxIn (..), TxOut (..), TxWitness (..) )
 import Data.Quantity
     ( Quantity (..) )
+import Data.Text
+    ( Text )
 import Data.Word
     ( Word16, Word8 )

@@ -82,7 +83,7 @@ data TransactionLayer t = TransactionLayer

     , validateSelection
       :: CoinSelection
-      -> Either ErrValidateSelection ()
+      -> Either Text ()
       -- ^ Validate coin selection
     }

@@ -95,13 +96,6 @@ data ErrMkStdTx
     -- or when transaction has more than 255 inputs or outputs (Jormungandr)
     deriving (Eq, Show)

-data ErrValidateSelection
-    = ErrInvalidTxOutAmount
-    -- ^ transaction with 0 amount is tried
-    | ErrExceededInpsOrOuts
-    -- ^ transaction has more than 255 inputs or outputs
-    deriving (Eq, Show)
-

 -- | Backend-specific variables used by 'estimateMaxNumberOfInputsBase'.
 data EstimateMaxNumberOfInputsParams t = EstimateMaxNumberOfInputsParams
diff --git a/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs b/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
index 268fc4f9..fb5d15d6 100644
--- a/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
+++ b/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
@@ -40,11 +40,7 @@ import Cardano.Wallet.Primitive.Types
     , txId
     )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..)
-    , ErrValidateSelection (..)
-    , TransactionLayer (..)
-    , estimateMaxNumberOfInputsBase
-    )
+    ( ErrMkStdTx (..), TransactionLayer (..), estimateMaxNumberOfInputsBase )
 import Control.Monad
     ( forM, when )
 import Data.ByteString
@@ -94,7 +90,7 @@ newTransactionLayer = TransactionLayer

     , validateSelection = \(CoinSelection _ outs _) -> do
         when (any (\ (TxOut _ c) -> c == Coin 0) outs)
-            $ Left ErrInvalidTxOutAmount
+            $ Left "Contains one or more outputs with value 0"
     }
   where
     mkWitness
diff --git a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
index d45e8029..a7f4ebc4 100644
--- a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
+++ b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
@@ -24,11 +24,7 @@ import Cardano.Wallet.Primitive.CoinSelection
 import Cardano.Wallet.Primitive.Types
     ( Hash (..), TxOut (..), TxWitness (..), txId )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..)
-    , ErrValidateSelection (..)
-    , TransactionLayer (..)
-    , estimateMaxNumberOfInputsBase
-    )
+    ( ErrMkStdTx (..), TransactionLayer (..), estimateMaxNumberOfInputsBase )
 import Control.Arrow
     ( second )
 import Control.Monad
@@ -63,8 +59,11 @@ newTransactionLayer (Hash block0) = TransactionLayer
         estimateMaxNumberOfInputsBase @t Binary.estimateMaxNumberOfInputsParams

     , validateSelection = \(CoinSelection inps outs _) -> do
-        when (length inps > 255 || length outs > 255)
-            $ Left ErrExceededInpsOrOuts
+        let validationErr = Left
+        when (length inps > 255) $
+            validationErr "Number of inputs exceeded 255."
+        when (length outs > 255) $
+            validationErr "Number of outputs exceeded 255."
     }
   where
     sign
```

</details>